File: project_euler/Problem062/Problem062.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cmath>
using namespace std;

class Problem062 {
public:
    void solution() {
        int target = 5;
        unordered_map<long long, vector<long long>> cubes(0);
        int index = 5;
        int pivot = 0;
        while (true) {
            long long value = (long long) pow(index, 3);
            long long digits = get_digits(value);
            int count = (int) log10(value);
            if (pivot != count) {
                cubes.clear();
                pivot = count;
            } else {
                if (cubes.find(digits) != cubes.end() && (int) cubes[digits].size() == target - 1) {
                    cout << cubes[digits][0] << endl;
                    return;
                }
            }
            if (cubes.find(digits) == cubes.end()) {
                cubes.insert(pair<long long, vector<long long>>(digits, vector<long long>(0)));
            }
            cubes[digits].emplace_back(value);
            ++index;
        }
    }

private:
    long long get_digits(long long value) {
        long long result = 0L;
        while (value) {
            result += (long long) pow(10, value % 10);
            value /= 10;
        }
        return result;
    }
};

File: project_euler/Problem062/Problem062.js
function Problem062() {
    let target = 5
    let cubes = new Map()
    let index = 5
    let pivot = 0
    while (true) {
        let value = Math.pow(index, 3)
        let digits = getDigits(value)
        let count = Math.floor(Math.log10(value))
        if (pivot !== count) {
            cubes.clear()
            pivot = count
        } else {
            if (cubes.has(digits) && cubes.get(digits).length === target - 1) {
                console.log(cubes.get(digits)[0])
                return
            }
        }
        if (!cubes.has(digits)) {
            cubes.set(digits, [])
        }
        cubes.get(digits).push(value)
        ++index
    }
}

function getDigits(value) {
    let result = 0
    while (value) {
        result += Math.pow(10, value % 10)
        value = Math.floor(value / 10)
    }
    return result
}

module.exports = {
    Problem062 : Problem062
};
